device: Optimize device_get_binding search efficiency

rruuaanng opened the pull request:
Optimize 'device_get_binding' function search so that when a matching device is found, it returns NULL if it is not ready.

Another solution.
```
	STRUCT_SECTION_FOREACH(device, dev) {
		if (dev->name == name ||strcmp(name, dev->name) == 0) {
			return z_impl_device_is_ready(dev) ? dev : NULL;
		}
	}
```
But. This simplifies it to a single loop and preserves the original requirement (I'm not sure if this solution will be unusual in large-scale deployments)

cfriedt commented:
I kind of prefer the other solution mentioned above, but I think there should be additional parenthesis because this is a core part of the OS and should abide by certain MISRA rules.

```cpp
STRUCT_SECTION_FOREACH(device, dev) {
		if ((dev->name == name) || (strcmp(name, dev->name) == 0)) {
			return z_impl_device_is_ready(dev) ? dev : NULL;
		}
	}
```

Another thought, and maybe I'm mistaken, but IIRC Zephyr rarely ever does an actual string comparison on the device name anymore. I think it would only be needed for devices that are not included in Devicetree, and potentially for LLEXT modules.

It's hard to say how much of an optimization this is as well without having exact numbers. I also suspect that since most devices are initialized in a predetermined order, we might be optimizing for the cold path.

gmarull commented:
> I kind of prefer the other solution mentioned above, but I think there should be additional parenthesis because this is a core part of the OS and should abide by certain MISRA rules.
> 
> ```c++
> STRUCT_SECTION_FOREACH(device, dev) {
> 		if ((dev->name == name) || (strcmp(name, dev->name) == 0)) {
> 			return z_impl_device_is_ready(dev) ? dev : NULL;
> 		}
> 	}
> ```
> 

+1

rettichschnidi commented:
Once merged, this would close #41665

rruuaanng committed:
device: Optimize device_get_binding search efficiency

Optimize 'device_get_binding' function search so that
when a matching device is found, it returns NULL if
it is not ready.

Signed-off-by: James Roy <rruuaanng@outlook.com>

rruuaanng commented:
> It's hard to say how much of an optimization this is as well without having exact numbers. I also suspect that since most devices are initialized in a predetermined order, we might be optimizing for the cold path.

The key point of optimization is that if the device is not ready, but it already exists, it will run the second loop after the first loop exits, and repeat the search and then return. After optimization, it is obvious that if it is not ready, it returns NULL and no more search(because it is very likely that it is still not ready in the second loop, which makes no sense. Maybe the speed may not be improved, in the extreme case, that is, in the second round of search, it is still not prepared and runs the second search in full.)

> we might be optimizing for the cold path.

You are right, the patch includes cold path optimization :)
